rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
+
+NS storage keeps different notification statuses for every user. Most Kibana deployments have [from 20 to 90 users](https://elastic.slack.com/archives/C9097ABGC/p1597162974399900),
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
+  notification_id: string;
+  recipient_id: string;
+  read_status: boolean;
+  pinned_status: boolean;
+}
+``` 
+This option consumes less storage space but might lead to expensive read operations due to an additional JOIN call to 
+search for notification state objects. We don’t expect Kibana to read more than 10-20 notifications at once, so the overhead might be acceptable.
+- Create a copy for every notification in a user-specific list of notifications. It’s [the recommended way to store data](https://www.elastic.co/guide/en/elasticsearch/reference/current/joining-queries.html),
+but it leads to duplicating notification messages. 
+
+Stale notifications and their states must be removed from the storage. There should be a dedicated task performing
+look up through the storage to remove notifications with *expired_at* > the current timestamp.
+
+Data are stored in the hidden system index to prevent accidental access by a random user.
+We don’t expect notifications to contain sensitive information.
+They must be stored in an encrypted form otherwise. 
+Kibana can query user-specific notifications by *recipent_id*.
+
+Alternative implementation might be to use SO with OLS([#17888](https://github.com/elastic/kibana/issues/17888)),
+but it might be overhead as we don't need any advanced SO functionality like export/import.
+Also, SO search functionality is quite limited and doesn’t allow using complex search queries. 
+
+### RemoteRepository
+Cloud plugin registered RemoteRepository as a strategy Kibana Notification Service for creation, storage,
+and retrieval notifications from the remote Unified Notification Service. 
+It’s still not clear what functionality is supported by UNS, but we expect it to be:
+- Create a notification for a particular recipient, a group of recipients. 
+- Read a list notification for a particular recipient.
+- Filter notification list.
+- Change a notification state.
+
+Kibana-specific entities shouldn’t leak to UNS: 
+- Kibana sends message translations for all the supported locales.
+- Kibana-specific groups of recipients unfolded to a list of *recipient_id*.
+- *recipient_id* is set to *elastic_id*.
```

Comment:
What is the distinction between a `recipient_id` and an `elastic_id`? Will _all_ users on ESS have a unique `elastic_id`?

Reply:
> What is the distinction between a recipient_id and an elastic_id? 

I *suspect* it will be `elastic_id` as long as we need to unambiguously identify a user across the Stack and Cloud.
In any case, Notification plugin relies on:
- `recipient_id` passed in from outside when creating the notification. So `recipient_id` form is defined  by `Security` plugin eventually.
- getting `user id`  from  Security plugin when requesting notifications from Kibana UI.

I'd refer to Security team whether:
- `Will all users on ESS have a unique elastic_id?`
- Will `elastic_id` be created for on-prem Kibana users or `user id` will be in another format?

Reply:
I don't know the answer here, but I do think it's important to not overlook. As far as I'm aware, we don't have precedent to rely on when figuring out how to translate from "Cloud users" to "Stack users".